### PR TITLE
CMake: Sanitize IIO feature list

### PR DESCRIPTION
**Pull request by pcercuei**
The options may not just be ON or OFF, they could be YES/NO, or 1/0.

Handle these cases with a new toggle_iio_feature() macro.

@rgetz

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -458,6 +458,15 @@ if(NOT SKIP_INSTALL_ALL)
 	install(FILES ${LIBIIO_PC} DESTINATION "${INSTALL_PKGCONFIG_DIR}")
 endif()
 
+macro(toggle_iio_feature opt str)
+	if (${opt})
+		set(_iio_feature_${opt} ON)
+	else()
+		set(_iio_feature_${opt} OFF)
+	endif()
+	list(APPEND IIO_FEATURES_${_iio_feature_${opt}} ${str})
```
[the remark this answers is not in the document]
Comment (pcercuei) [reply CMakeLists.txt:467]: I guess :)

**Comment by rgetz** [comment]
I think this gives:

True if the constant is 1, ON, YES, TRUE, Y, or a non-zero number (including floating point numbers). 
False if the constant is 0, OFF, NO, FALSE, N, IGNORE, NOTFOUND, the empty string, or ends in the suffix -NOTFOUND.

Which should cover what we want...


And - not a show stopper, but this does make it a little harder to build the test/utils by themselves based on the installed libiio.

```
mkdir build && cd build
cmake ../tests
make
```

now fails, since it can't find iio-config.h - an easy fix
now fails, since it doesn't know where the pthread lib is (it's set in the top level CMake). - also an easy fix
and will fail  with this patch, since it calls a macro that it doesn't know of (it's defined in the top level CMake, rather than being included from the cmake dir).

-Robin

**Comment by rgetz** [comment]
Just a quick check of the CI:

LinuxBuilds ubuntu_18_04:
```
-- Features enabled : xml network dns-sd avahi ipv6 local local-mmap usb utils iiod iiod-serial iiod-aio iiod-usb man man-utils python-bindings
-- Features disabled: zstd bonjour serial hwmon examples iiod-systemd iiod-sysvinit iiod-upstart c#-bindings
```

LinuxBuilds ubuntu_20_04:
```
-- Features enabled : xml network dns-sd avahi ipv6 local local-mmap usb utils iiod iiod-serial iiod-aio iiod-usb man man-utils python-bindings
-- Features disabled: zstd bonjour serial hwmon examples iiod-systemd iiod-sysvinit iiod-upstart c#-bindings
```

LinuxBuilds ubuntu_22_04:
```
-- Features enabled : xml network dns-sd avahi ipv6 local local-mmap usb utils iiod iiod-serial iiod-aio iiod-usb man man-utils python-bindings
-- Features disabled: zstd bonjour serial hwmon examples iiod-systemd iiod-sysvinit iiod-upstart c#-bindings
```

LinuxBuilds fedora34:
```
-- Features enabled : xml network dns-sd avahi ipv6 local local-mmap usb utils iiod iiod-serial iiod-aio iiod-usb man man-utils python-bindings
-- Features disabled: zstd bonjour serial hwmon examples iiod-systemd iiod-sysvinit iiod-upstart c#-bindings
```

WindowsBuilds VS2022:
```
-- Features enabled : xml network dns-sd ipv6 serial usb utils man-utils python-bindings c#-bindings
-- Features disabled: zstd avahi bonjour local local-mmap hwmon examples iiod man
```

WindowsBuilds VS2019:
```
-- Features enabled : xml network dns-sd ipv6 serial usb utils man-utils python-bindings c#-bindings
-- Features disabled: zstd avahi bonjour local local-mmap hwmon examples iiod man
```

macOSBuilds macOS_11:
```
-- Features enabled : xml network dns-sd bonjour ipv6 usb utils examples man-utils python-bindings
-- Features disabled: zstd avahi serial local local-mmap hwmon iiod man c#-bindings
```

macOSBuilds macOS_12:
```
-- Features enabled : xml network dns-sd bonjour ipv6 usb utils examples man-utils python-bindings
-- Features disabled: zstd avahi serial local local-mmap hwmon iiod man c#-bindings
```

ARMBuilds ubuntu-ppc64le:
```
-- Features enabled : xml network dns-sd avahi ipv6 serial local local-mmap hwmon usb utils examples iiod iiod-serial iiod-aio iiod-usb man-utils python-bindings
-- Features disabled: zstd bonjour iiod-systemd iiod-sysvinit iiod-upstart man c#-bindings
```

ARMBuilds ubuntu-x390x:
```
-- Features enabled : xml network dns-sd avahi ipv6 serial local local-mmap hwmon usb utils examples iiod iiod-serial iiod-aio iiod-usb man-utils python-bindings
-- Features disabled: zstd bonjour iiod-systemd iiod-sysvinit iiod-upstart man c#-bindings
```

ARMBuilds ubuntu-arm64v8:
```
-- Features enabled : xml network dns-sd avahi ipv6 serial local local-mmap hwmon usb utils examples iiod iiod-serial iiod-aio iiod-usb man-utils python-bindings
-- Features disabled: zstd bonjour iiod-systemd iiod-sysvinit iiod-upstart man c#-bindings
```

ARMBuilds ubuntu-arm32v7:
```
-- Features enabled : xml network dns-sd avahi ipv6 serial local local-mmap hwmon usb utils examples iiod iiod-serial iiod-aio iiod-usb man-utils python-bindings
-- Features disabled: zstd bonjour iiod-systemd iiod-sysvinit iiod-upstart man c#-bindings
```

From this - we can see a couple things (that are maybe outside the scope of this patch):
 - there is no flag/print for mdns (the mdns solution we use for Windows).
 - zstd never gets turned on, on any target?
 - serial & hwmon don't get turned on for many targets? 
 - doc is missing.

cmake warns on some platforms:
CMake Deprecation Warning at tests/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

since:
```
bindings/csharp/CMakeLists.txt:cmake_minimum_required(VERSION 2.8.10)
bindings/python/CMakeLists.txt:cmake_minimum_required(VERSION 2.8.10)
CMakeLists.txt:cmake_minimum_required(VERSION 2.8.10)
examples/CMakeLists.txt:cmake_minimum_required(VERSION 2.8.10)
iiod/CMakeLists.txt:cmake_minimum_required(VERSION 2.8.10)
tests/CMakeLists.txt:cmake_minimum_required(VERSION 2.8.10)
```
is less of an issue know that no one should be using the old Linaro releases.

And the various Doxyfile.in files should be regenerated, since we get various warnings...
```
Generating API documentation with Doxygen
warning: Tag 'TCL_SUBST' at line 168 of file '/__w/1/s/build/Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 871 of file '/__w/1/s/build/Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'PERL_PATH' at line 1558 of file '/__w/1/s/build/Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'MSCGEN_PATH' at line 1580 of file '/__w/1/s/build/Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
Generating C# documentation with Doxygen
warning: Tag 'TCL_SUBST' at line 168 of file '/__w/1/s/build/Doxyfile_csharp' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 871 of file '/__w/1/s/build/Doxyfile_csharp' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'PERL_PATH' at line 1558 of file '/__w/1/s/build/Doxyfile_csharp' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
warning: Tag 'MSCGEN_PATH' at line 1580 of file '/__w/1/s/build/Doxyfile_csharp' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
```

rgetz@brain:~/github/libiio$ doxygen  -u Doxyfile.in^C
rgetz@brain:~/github/libiio$ git diff Doxyfile.in | diffstat
 Doxyfile.in | 1135 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------
 1 file changed, 947 insertions(+), 188 deletions(-)

Someone should go through and make sure it doesn't break anything.

**Comment by pcercuei** [comment]
> And - not a show stopper, but this does make it a little harder to build the test/utils by themselves based on the installed libiio.
> 
> ```
> mkdir build && cd build
> cmake ../tests
> make
> ```
> 
> now fails, since it can't find iio-config.h - an easy fix now fails, since it doesn't know where the pthread lib is (it's set in the top level CMake). - also an easy fix and will fail with this patch, since it calls a macro that it doesn't know of (it's defined in the top level CMake, rather than being included from the cmake dir).

But this has never been supported, and I don't see why it should.

**Comment by rgetz** [comment]
> > And - not a show stopper, but this does make it a little harder to build the test/utils by themselves based on the installed libiio.
> > [snip]
> 
> But this has never been supported, and I don't see why it should.

When you want to build a new version of the utilities with an old version of the library; as well as just keeping the utilities separate from the lib... I have done this alot in the past.

Like I said - not a show stopper, and shouldn't have anything to do with this PR directly.

**Comment by pcercuei** [comment]
Ok. Well I did not know this was a use case so it broke at some point.
This PR breaks it even more, but since it is already the case in master, I guess we can ignore it for now, yes.

**Comment by rgetz** [comment]
so, I think the only outstanding thing is : 

`list(APPEND IIO_FEATURES_${WITH_DOC} doc)`

wasn't replaced - it use to print out "doc", when doc was enabled/disabled, and now it does not. Was that on purpose?

**Comment by pcercuei** [comment]
Oops.

**Review by rgetz — APPROVED** [review]
Looks good to me - thanks for updating.

**Review by nunojsa — APPROVED** [review]

**Pull request merged**